fdbrpc/DDSketch.h

```diff
@@ -0,0 +1,190 @@
+#include <vector>
+#include <algorithm>
+#include <cmath>
+#include "flow/Error.h"
+
+#if (!(__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__))
+#error Do not support non-little-endian systems
+#endif
+
+#define FASTLOG 1
+// This significantly reduces the time for log
+
+#if FASTLOG
+struct fastLogger {
+	// Basically, the goal is to compute log(x)/log(r).
+	// For double, it is represented as 2^e(1+s) (0<=s<1), so our goal becomes e*log(2)/log(r)*log(1+s),
+	// and we approximate log(1+s) with a cubic function. See more details on DataDog's paper, or
+	// CubicallyInterpolatedMapping.java in https://github.com/DataDog/sketches-java/
+	constexpr const static double correctingFactor = 1.00988652862227438516; // = 7 / (10 * log(2));
+	constexpr static const double A = 6.0 / 35.0, B = -3.0 / 5.0, C = 10.0 / 7.0;
+	static const int SIGNIFICAND_WIDTH = 53;
+	static const uint64_t SIGNIFICAND_MASK = 0x000fffffffffffffuLL;
+	static const uint64_t EXPONENT_MASK = 0x7FF0000000000000uLL;
+	constexpr static const int EXPONENT_SHIFT = SIGNIFICAND_WIDTH - 1;
+	static const int EXPONENT_BIAS = 1023;
+	static const uint64_t ONE = 0x3ff0000000000000uLL;
+	static inline uint64_t doubleToLongBits(double x) {
+		uint64_t u;
+		uint64_t* pu = reinterpret_cast<uint64_t*>(&x);
+		u = *pu;
+		return u;
+	}
+	static inline double longBitsToDouble(uint64_t x) {
+		double u;
+		double* pu = reinterpret_cast<double*>(&x);
+		u = *pu;
+		return u;
+	}
+	static int getExponent(uint64_t longBits) {
+		return (int)((longBits & EXPONENT_MASK) >> EXPONENT_SHIFT) - EXPONENT_BIAS;
+	}
+
+	static double getSignificandPlusOne(uint64_t longBits) {
+		return longBitsToDouble((longBits & SIGNIFICAND_MASK) | ONE);
+	}
+	static double fastlog(double value) {
+		uint64_t longBits = doubleToLongBits(value);
+		double s = getSignificandPlusOne(longBits) - 1;
+		double e = getExponent(longBits);
+		return ((A * s + B) * s + C) * s + e;
+	}
+};
+#endif
+
+// A DDSketch for non-negative numbers
+// (those < 10^-18 are treated as 0, and huge numbers (>>10^18) may degrade the performance)
+template <class T>
+class DDSketch {
+public:
+	explicit DDSketch(double errorGuarantee = 0.1)
+	  : errorGuarantee(errorGuarantee), gamma((1.0 + errorGuarantee) / (1.0 - errorGuarantee)), logGamma(log(gamma)),
+	    multiplier(FASTLOG ? fastLogger::correctingFactor * log(2) / log(1 + 2 * errorGuarantee / (1 - errorGuarantee))
+	                       : 1),
+	    offset((FASTLOG ? fastLogger::fastlog(1 / EPS) * multiplier : log(1 / EPS)) + 5), populationSize(0),
+	    zeroPopulationSize(0), _min(T()), _max(T()), sum(T()) {
+
+		buckets.resize(2 * offset, 0);
+#if FASTLOG
+		ASSERT(sizeof(double) == sizeof(uint64_t));
+#endif
+	}
+
+	DDSketch<T>& addSample(T sample) {
+		// Call it addSample for now, while it is not a sample anymore
+		if (!populationSize) _min = _max = sample;
+
+		if (sample <= EPS) {
+			zeroPopulationSize++;
+		} else {
+#if FASTLOG
+			int index = ceil(fastLogger::fastlog(sample) * multiplier);
```

Comment:
Definitely add some tests for this. I've messed up the floor/ceil index logic a bunch of times and won't swear that I can verify this implementation by reading code alone.

Reply:
Well this definitely worths more investigation. The paper uses the ceil, and I noticed the Java implementation used floor. In my (private) unit test, I tried both, and the ceil is far more accurate. I will dig more and report back.

Reply:
I checked the paper and communicated with the authors. The Java reference implementation is slightly different from the paper, and my implementation is consistent with the paper. I did not see anyone is better than the other so I will stick with the current state.